numpy/_core/src/multiarray/stringdtype/casts.c

```diff
@@ -1503,13 +1507,258 @@ STRING_TO_CFLOAT_CAST(clongdouble, l, longdouble)
+  1,2,1,1,1,1,1,1,1,2,1,1,1,1,1,1,1,1,1,1,1,1,1,3,1,3,1,1,1,1,1,1, // s5..s6
+  1,3,1,1,1,1,1,3,1,3,1,1,1,1,1,1,1,3,1,1,1,1,1,1,1,1,1,1,1,1,1,1, // s7..s8
+};
+
+static uint32_t inline
+utf8_decode(uint32_t* state, uint32_t* codep, uint32_t byte) {
+    uint32_t type = utf8d[byte];
+
+    *codep = (*state != UTF8_ACCEPT) ?
+            (byte & 0x3fu) | (*codep << 6) :
+            (0xff >> type) & (byte);
+
+    *state = utf8d[256 + *state*16 + type];
+    return *state;
+}
+
+static int
+utf8_buffer_size(uint8_t *s, size_t *encoded_size_in_bytes, size_t max_bytes) {
+    uint32_t codepoint;
+    uint32_t state = 0;
+    size_t num_bytes = 0;
+    *encoded_size_in_bytes = 0;
+
+    // ignore trailing nulls
+    while (max_bytes > 0 && s[max_bytes - 1] == 0) {
+        max_bytes--;
+    }
+
+    if (max_bytes == 0) {
+        return UTF8_ACCEPT;
+    }
+
+    for (; num_bytes < max_bytes; ++s)
+    {
+        utf8_decode(&state, &codepoint, *s);
+        if (state == UTF8_REJECT)
+        {
+            return state;
+        }
+        else if(state == UTF8_ACCEPT)
+        {
+            *encoded_size_in_bytes += num_utf8_bytes_for_codepoint(codepoint);
+        }
+        num_bytes += 1;
+    }
+
+    return state != UTF8_ACCEPT;
+}
+
+static int
+void_to_string(PyArrayMethod_Context *context, char *const data[],
+               npy_intp const dimensions[], npy_intp const strides[],
+               NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_Descr **descrs = context->descriptors;
+    StringDTypeObject *descr = (StringDTypeObject *)descrs[1];
+
+    npy_string_allocator *allocator = NpyString_acquire_allocator(descr);
+
+    long max_in_size = descrs[0]->elsize;
+
+    npy_intp N = dimensions[0];
+    unsigned char *in = (unsigned char *)data[0];
+    char *out = data[1];
+
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1];
+
+    while(N--) {
+        size_t out_num_bytes = 0;
+        if (utf8_buffer_size(in, &out_num_bytes, max_in_size) == UTF8_REJECT) {
+            npy_gil_error(PyExc_TypeError,
+                          "Invalid UTF-8 bytes found, cannot convert to UTF-8");
+            goto fail;
+        }
+        npy_packed_static_string *out_pss = (npy_packed_static_string *)out;
+        if (NpyString_free(out_pss, allocator) < 0) {
+            npy_gil_error(PyExc_MemoryError,
+                      "Failed to deallocate string in unicode to string cast");
+            goto fail;
+        }
+        if (NpyString_newemptysize(out_num_bytes, out_pss, allocator) < 0) {
+            npy_gil_error(PyExc_MemoryError,
+                      "Failed to allocate string in unicode to string cast");
+            goto fail;
+        }
+        npy_static_string out_ss = {0, NULL};
+        int is_null = NpyString_load(allocator, out_pss, &out_ss);
+        if (is_null == -1) {
+            npy_gil_error(PyExc_MemoryError,
+                      "Failed to load string in unicode to string cast");
```

Comment:
Little copy paste error here, I guess it probably cannot really fail.  Do you even need this part?

Reply:
Thanks! It can fail if there is memory corruption or a programming error. I'd rather leave the error checking so people get an error in Python right after the issue occurs instead of weirder behavior later.